src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/SignatureAlgorithmFactory.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\SecurityBundle\DependencyInjection\Security\Factory;
+
+use Jose\Component\Core\Algorithm as SignatureAlgorithm;
+use Jose\Component\Signature\Algorithm;
+use Symfony\Component\Security\Core\Exception\InvalidArgumentException;
+use Symfony\Component\Security\Http\AccessToken\Oidc\OidcTokenHandler;
+
+/**
+ * Creates a signature algorithm for {@see OidcTokenHandler}.
+ *
+ * @experimental
+ */
+final class SignatureAlgorithmFactory
```

Comment:
I think it'd make sense to put this factory in the JOSE package (can be done later though)

Reply:
/cc @Spomky WDYT?

Reply:
I'm not sure requiring the `web-token/jwt-framework` just to use this factory system (which is not the same approach than Symfony out of framework-bundle) is a good approach in this PR.

Reply:
Note that [AlgorithmManagerFactory](https://github.com/web-token/jwt-framework/blob/3.2.x/src/Component/Core/AlgorithmManagerFactory.php) is in the `web-token/jwt-core` package.

Reply:
Hi,

Thanks for sharing those elements I haven't seen earlier!

Considering that, should we force the user to require and use the WebToken bundle, in order to use those elements? For example:
```php
if (!ContainerBuilder::willBeAvailable('web-token/jwt-bundle', AlgorithmCompilerPass::class, ['symfony/security-bundle'])) {
```

IMO, requiring an external bundle in a Symfony bundle is vendor-locking and restrictive, and should be avoided. I would prefer requiring `web-token/jwt-signature-*` libraries only, which could be easily replaced later if necessary as it impacts less configuration and less files in the user-side Symfony project.

Also, using the AlgorithmManagerFactory + WebToken bundle would require to update the OidcTokenHandler DI: remove the signature object, inject the AlgorithmManagerFactory dependency + the signature protocol string, and create the signature object on runtime.

IMO, this won't respect the [LoD](https://en.wikipedia.org/wiki/Law_of_Demeter). I would prefer to keep the design and the DI simple by injecting the signature object directly.

WDYT?

Reply:
I mentioned the WebToken Bundle for illustrating how the alg manager factory is used. The bundle is not needed at all.

What could be done here:

* Define the [AlgorithmManager](https://github.com/web-token/jwt-framework/blob/3.2.x/src/Component/Core/AlgorithmManager.php) as a service (let say `security.access_token_handler.oidc_algorithm_manager`)
* Define ESxx algorithms as services with a specific tag (ex: `security.access_token_handler.signature_algorithm`)
* Define a compiler pass that will gather all services tagged with `security.access_token_handler.signature_algorithm` and add them to `security.access_token_handler.oidc_algorithm_manager` (the method to call is [`add`](https://github.com/web-token/jwt-framework/blob/3.2.x/src/Component/Core/AlgorithmManager.php#L61))
* Use `security.access_token_handler.oidc_algorithm_manager` and the `has`/`get` methods to get the algorithm instead of this class.
* Other algorithms will be automatically added just by tagging them.

Reply:
> IMO, this won't respect the LoD. I would prefer to keep the design and the DI simple by injecting the signature object directly.

OK, my solution above is not really inline with that as is will require the AlgorithmManager to be injected to get the right algorithm.